### PR TITLE
Rework usage / --help output

### LINKED ISSUE #4340: Rework usage / --help output
I think our hiding of plenty of useful options under `--list=hidden-options` is too obscure. I think we should unhide them all, printing the full options list on `--help`. We're being inconsistent with other projects, including hashcat, by not printing the full options list even on explicit `--help`. This can be confusing to new users, who wouldn't realize there are more options nor how to see those.

If we don't want to flood the screen with all the options when `john` is invoked with no options (not even with `--help`), then we should limit its default output to just a few lines like the current first 5 lines of output (which are: program name and build info, copyright, homepage, empty line, general command-line syntax) plus the suggestion to use `--help`, so maybe 6 or 7 lines total.

Yes, this will make jumbo not print even the option names that core does by default, but so what. Core's full options list fits on screen, and jumbo's doesn't (by far), which justifies the difference. We have a crucial and more confusing difference now: core's options list printed by default is complete, and jumbo's currently isn't - so let's avoid that one.

Related/historical: #1046

### PR DESCRIPTION
**Pull request by wsandin**
Additional details to be filled out. This pull request is currently WIP.

Fixes #4340

### CODE REVIEW

**Comment by ghost** [comment]
I like the new `./john` but I dislike the new `./john --help`. @wsandin , it is not about the PR. 

Solar, my two cents would be: do like git  works (eh, they use `man` for detailed information). Then, the way snap works.
```
snap --help            # 10-15 lines of info
snap login --help      # more 10-15 lines of detailed info

git --help
git push --help  # man GIT-PUSH(1)
```

```
$ ../run/john --help | wc -l
114
```

```
$ git --help | wc -l
45
```

**Comment by ghost** [comment]
**[edited]** Thinking out loud:
```
john opencl  --help
john prince  --help
john markov  --help
And so on (groups that make sense).
```

Using **any** build, I can:
```
$ john ztex  --help
### Resource is DISABLED ## (OR NOT, of course)
Uses a ztex board [...] cost effective [...] watt per dolar [...] some FPGAs [...]
```
Etc.
```
$ john opencl  --help
Uses a GPU [...] efficient for a wide range of hashes [...] cloud providers [...]
```

File: src/john.c
Comment (magnumripper) [inline src/john.c]: Whitespace violation, we indent with tabs and align with spaces (I'll not look for or comment on any further instances)

File: src/options.c
Comment (magnumripper) [inline src/options.c]: Here's a good chance to bump to "1996-2020"

File: src/options.c
Comment (magnumripper) [inline src/options.c]: Is there some rationale for adding 50 spaces here?

File: src/options.c
Comment (magnumripper) [inline src/options.c:368]: On a side note we should rename this option to `--wordlist-mem-file-size` or something else starting with `--wordlist-` (such as `--wordlist-preload-threshold`)

We'd do so in another PR though, the deprecated old option name would need to be handled for a while.

File: src/options.c
Comment (magnumripper) [inline src/options.c]: ...and ilkewise these should be `--wordlist-dupe-suppression` and `--wordlist-loopback`.

PRINCE option is great with tab completion in that you can quickly find all options by saying `--prince` and hit tab twice. Renaming the mentioned options would give the same advantage.

File: src/options.c
Comment (magnumripper) [inline src/options.c]: I'd move `--config` to at least after `JOHN_USAGE_REGEX` but probably much more down. Perhaps such shuffling (I can think of more) should be a separate issue and PR though.

Arguably we could also opt to list all options in alphabetical order.

File: src/options.c
Comment (magnumripper) [inline src/options.c]: --incremental-charcount should go right after `--incremental`

File: src/options.c
Comment (magnumripper) [inline src/options.c]: These four should be right after `--test`

File: src/options.c
Comment (magnumripper) [inline src/options.c]: `--rules-skip-nop` should go right after `--rules`

File: src/options.c
Comment (magnumripper) [inline src/options.c]: We already listed `--format=NAME`. Now when we drop hidden options that one should be `--list=[NAME|CLASS][,...]`

File: src/options.c
Comment (magnumripper) [inline src/options.c]: `--mkv-stats` should go right after `--markov` (and is later subject to rename to `--markov-stats`)

File: src/options.c
Comment (magnumripper) [inline src/options.c]: These ought to go ~right after `--format`~, sorry, right after `--show`

File: src/options.c
Comment (magnumripper) [inline src/options.c]: These should go right after `--subsets`

**Review by magnumripper — COMMENTED** [review]
I had a look and added MHO

File: src/options.c
Comment (magnumripper) [reply src/options.c]: BTW I see we forgot to update this. Current valid classes are `dynamic, cpu, opencl, ztex, mask, omp, all, enabled, disabled`. So something more like
```
"--format=[NAME|CLASS]      force hash type. The supported formats can be seen\n" \
"                           with --list=formats and --list=subformats. Valid\n" \
"                           classes: dynamic, cpu, opencl, ztex, mask, omp, all,\n" \
"                           enabled, disabled.\n"
```

**Comment by wsandin** [comment]
Hi @magnumripper and @claudioandre-br,

Thanks for all the valuable feedback.

I'm sorry I forgot to post my message yesterday highlighting that this was just a Pull Request for me to ask for some feedback in regards the duplication of `--format`, and how we want to sort some of the commands. I'll indeed clear out any IDE misconfiguration clutter (such as the additional whitespaces, wrong indentation) and have a finecomb before asking for a final review again.

In regards to your feedback @claudioandre-br, I agree that it would be practical and congruent with a lot of other tools out there. Question is if we want to extend the scope of this issue to that, @solardiz? I'd be happy to take it on in that case, but It'd definitely helpful to get this cleared out before I start making more progress on the structure of the john man page.

**Comment by magnumripper** [comment]
No problem - I just took a quick look and made some reflections (where some was out of scope but should be considered for later). Also, I agree with Claudio but I think that too is out of scope right now.

**Comment by solardiz** [comment]
> In regards to your feedback @claudioandre-br, I agree that it would be practical and congruent with a lot of other tools out there. Question is if we want to extend the scope of this issue to that, @solardiz?

I don't want to extend the scope. I appreciate Claudio's suggestion, but I think he didn't want that either as he wrote "it is not about the PR". We might revisit that separately. Thanks!

File: src/options.c
Comment (magnumripper) [reply src/options.c]: I'll refrain from changing this in #4395 to avoid merge problems. Please make my suggested change above!

File: src/john.c
Comment (wsandin) [reply src/john.c]: Got it! Thanks for the heads-up. I've reconfigured my editor for this project now. :+1:
I take the liberty in resolving this thread.

File: src/options.c
Comment (wsandin) [reply src/options.c]: Absolutely none, again thanks for the heads up and sorry about my IDE clutter. I'm normally a vim user, but thought an IDE could be helpful now that I'm trying to work on my C programming.

File: src/options.c
Comment (wsandin) [reply src/options.c]: Applied.

File: src/options.c
Comment (wsandin) [reply src/options.c]: Applied.

File: src/options.c
Comment (wsandin) [reply src/options.c]: Applied.

File: src/options.c
Comment (wsandin) [reply src/options.c]: Applied.

File: src/options.c
Comment (wsandin) [reply src/options.c]: Perfect, I applied the change to the `--format=` macro. However, it was unclear to me whether I should make any changes to `--list`?

Comment (magnumripper) [reply src/options.c]: No that was a typo of mine. I meant `--format=[NAME|CLASS][,...]`

Comment (magnumripper) [reply src/options.c]: Hmm so maybe I also meant to include `[,...]` like below but we're one character short so it'll just be `[,..]`
```
"--format=[NAME|CLASS][,..] force hash type. The supported formats can be seen\n" \
"                           with --list=formats and --list=subformats. Valid\n" \
"                           classes: dynamic, cpu, opencl, ztex, mask, omp, all,\n" \
"                           enabled, disabled.\n"
```

File: src/options.c
Comment (wsandin) [reply src/options.c]: Applied.

File: doc/OPTIONS
Comment (magnumripper) [inline doc/OPTIONS]: The default max. size file for this to happen is shown ~**in**~ among the --help options.

File: src/options.c
Comment (wsandin) [reply src/options.c:368]: I created https://github.com/openwall/john/issues/4399, and happy to take it on.

**Comment by wsandin** [comment]
I'll have a look at getting the shell completions to work now.

File: src/options.c
Comment (magnumripper) [inline src/options.c]: Not sure from my current view (kinda confusing wall of strings and macros) but I think line 444 shouldn't end with a `\` although in this incarnation the blank line hides that problem.

Comment (wsandin) [reply src/options.c]: Great catch! You're indeed right that it's quite confusing to me. But I hope to acquaint myself with the codebase by working on issues like this one. :-)

I just pushed a fix. There's a few noisy commits now, so maybe I should squash the commits. I can't find any option here in GitHub for the pull request to squash before merge. Maybe that's on the reviewer/merger's side?

File: src/options.c
Comment (wsandin) [reply src/options.c]: I covered these in #4399 as well.

File: src/options.c
Comment (wsandin) [reply src/options.c]: I put it just above logging and verbosity.

File: src/options.c
Comment (wsandin) [reply src/options.c]: Applied.

File: src/options.c
Comment (wsandin) [reply src/options.c]: Applied.

File: src/options.c
Comment (wsandin) [reply src/options.c]: Applied.

File: doc/OPTIONS
Comment (wsandin) [reply doc/OPTIONS]: Thanks, just fixed.

**Comment by magnumripper** [comment]
> I'll have a look at getting the shell completions to work now.

They were already lagging behind a bit, please search for open issues. If you can get them 100% you'll be my hero for quite some time. For example IIRC, they don't yet grok options with "parents", like `--single` vs. `--single-seed`

File: src/options.c
Comment (magnumripper) [reply src/options.c]: We can squash when merging but I think Solar mentioned recently it might screw up some metadata. As long as we're not very near merging, please squash at will!

**Comment by magnumripper** [comment]
> > I'll have a look at getting the shell completions to work now.
> 
> They were already lagging behind a bit, please search for open issues. If you can get them 100% you'll be my hero for quite some time. For example IIRC, they don't yet grok options with "parents", like `--single` vs. `--single-seed`

...and consequently perhaps even shell completion is out of scope for this issue. By all means fix what you can, but if things get hairy or scary, just open new issues and leave them out!

File: src/options.c
Comment (magnumripper) [inline src/options.c]: Line 408 here, belongs to `--test` and not `--test-full`

**Review by magnumripper — CHANGES_REQUESTED** [review]
Here's a bug now:
```
$ ../run/john test.in
John the Ripper 1.9.0-jumbo-1+bleeding-6f24296d0 2020-10-04 22:21:33 +0100 OMP [darwin19.6.0 64-bit x86_64 AVX2 AC]
Copyright (c) 1996-2020 by Solar Designer and others
Homepage: https://www.openwall.com/john/

Usage: john [OPTIONS] [PASSWORD-FILES]
Use --help to list all available options.

Warning: detected hash type "Office", but the string is also recognized as "office-opencl"
(...)
```
The usage blurb shouldn't be printed in this situation.

**Review by magnumripper — CHANGES_REQUESTED** [review]
Here's a related problem:
```
$ ../run/john --list=formats --format=nt*
John the Ripper 1.9.0-jumbo-1+bleeding-6f24296d0 2020-10-04 22:21:33 +0100 OMP [darwin19.6.0 64-bit x86_64 AVX2 AC]
Copyright (c) 1996-2020 by Solar Designer and others
Homepage: https://www.openwall.com/john/

Usage: john [OPTIONS] [PASSWORD-FILES]
Use --help to list all available options.

NT, NT-opencl, ntlmv2-opencl
```
Our Test Suite (and other things) depend on neither banner or usage hint printed here, only the requested information (ie. the last line here). **Edit:** recent versions add "3 formats" after it, but that's on stderr so doesn't disturb the tools.

File: src/options.c
Comment (wsandin) [reply src/options.c]: Fixed.

File: src/john.c
Comment (magnumripper) [inline src/john.c]: I just realized this call to `john_register_all()` is obsolete (actually ever since 2015 in 9e3bb759)! Just remove it.

File: src/options.c
Comment (magnumripper) [inline src/options.c]: 469-470 breaks MPI: You must not exit unless main node will also exit!

Simply doing this would work fine though (not sure if it's worth its own function at all though?):
```c
{
	if (john_main_process)
	    printf(JOHN_BANNER, name);
}
```
*(edited after copy-paste error)*

**Review by magnumripper — CHANGES_REQUESTED** [review]

File: src/john.c
Comment (magnumripper) [reply src/john.c]: ...and as a result of that, you should actually move this whole thing from `john.init()` to `opt_init()`, which would make its `int show_usage` parameter obsolete as well.

**Comment by magnumripper** [comment]
> Here's a bug now:
> 
> ```
> $ ../run/john test.in
> John the Ripper 1.9.0-jumbo-1+bleeding-6f24296d0 2020-10-04 22:21:33 +0100 OMP [darwin19.6.0 64-bit x86_64 AVX2 AC]
> Copyright (c) 1996-2020 by Solar Designer and others
> Homepage: https://www.openwall.com/john/
> 
> Usage: john [OPTIONS] [PASSWORD-FILES]
> Use --help to list all available options.
> 
> Warning: detected hash type "Office", but the string is also recognized as "office-opencl"
> (...)
> ```
> 
> The usage blurb shouldn't be printed in this situation.

The banner (topmost three lines plus the blank) historically also wasn't printed for a crack session like the above, but perhaps we do want it to print now @solardiz? If so, we need to think a little to avoid foot guns. We can't print any such thing to stdout if `--list=` option was used, and also not likely if `--stdout` option was used, right? Any more?

That would mean we can't decide until after `opt_process()` was called from `opt_init()`.

**Comment by solardiz** [comment]
> The banner (topmost three lines plus the blank) historically also wasn't printed for a crack session like the above, but perhaps we do want it to print now @solardiz?

No, I think we still don't.

File: src/options.c
Comment (wsandin) [reply src/options.c]: Thanks @magnumripper. I agree that it might justify it's own function. I created it in order to follow convention, and avoid using brackets to encapsulate the `if (john_main_process)`, in `opt_init()`. I have now merged these into the function - happy to hear your feedback.

File: src/john.c
Comment (wsandin) [reply src/john.c]: Cool. I removed all references to `john_register_all()`, and moved the conditionals from john_init to `opt_init()`

Comment (magnumripper) [reply src/john.c]: > I removed all references to `john_register_all()`

Oh, that will not work and wasn't what I said. I meant that one *call* to it only.

Comment (magnumripper) [reply src/john.c]: 3f6b666 is just like I meant, just revert 921eb19

Comment (wsandin) [reply src/john.c]: I just did. Thanks :) Unfortunately I'm down with a cold today, but hope to go through the rest in the next 1-2 days.

**Comment by wsandin** [comment]
Finally back on my feet. I should be able to complete this (as in finecombing and add shell completions where possible/applicable) by the weekend at least. Sorry the delay.

**Comment by magnumripper** [comment]
@wsandin you OK?

I'm holding off a few commits so not to clash with this one. Perhaps you could fix the one remaining problem here (the `exit(0)` in line 485) and we could merge this even though bash completion would be out of sync for a while.

**Comment by wsandin** [comment]
> @wsandin you OK?
> 
> I'm holding off a few commits so not to clash with this one. Perhaps you could fix the one remaining problem here (the `exit(0)` in line 485) and we could merge this even though bash completion would be out of sync for a while.

Hey @magnumripper. Apologies for my procastrination, I'll push that fix right away and then see if I can partially fix the completions, otherwise split that out to another issue. If that sounds good?

**Comment by wsandin** [comment]
~~I'm having a look at shell completions at the moment. I do get
Error: Cannot find John home. Invoke the program via full or relative pathname.
For example, /full/path/john or path/john, or set and use a shell alias.
When copying the `run/john` binary to `/usr/local/bin`. I don't see how that could be related to my changes though.~~

I got it to work with `./configure --system-wide` rather than `--prefix=`

**Comment by ghost** [comment]
One detail to fix before merging: the "unknown option" error message.

```
$ ../run/john --help
John the Ripper 1.9.0-jumbo-1+bleeding-4a413af 2020-10-28 23:54:06 +0300 OMP [linux-gnu 64-bit x86_64 AVX AC]
Copyright (c) 1996-2020 by Solar Designer and others
Homepage: https://www.openwall.com/john/

Usage: john [OPTIONS] [PASSWORD-FILES]
Use --help to list all available options.

--help                     print usage summary
[...]
--format=[NAME|CLASS][,..] force hash of type NAME. The supported formats can
                           be seen with --list=formats and --list=subformats. Valid
                           classes: dynamic, cpu, opencl, ztex, mask, omp, all,
                           enabled, disabled.
Unknown option: "--help"
```

**Comment by ghost** [comment]
@wsandin , IMO, if you create a new PR to handle shell completion, @magnumripper can approve the PR easier (he already tested the `--help` stuff).

**Comment by wsandin** [comment]
> @wsandin , IMO, if you create a new PR to handle shell completion, @magnumripper can approve the PR easier (he already tested the `--help` stuff).

That's a good point. I saw your message just now, I'll rebase these out of this branch then and move to a separate PR after a quick review tomorrow - want to go through the completions again.

File: src/options.c
Comment (magnumripper) [inline src/options.c]: While this doesn't hurt, it's not the correct fix and should be reverted. We handle --help separately. The problem you were trying to solve is you still don't `exit(0)` in the right places.

File: src/listconf.c
Comment (magnumripper) [inline src/listconf.c:404]: I think for compatibility with existing scripts/front-ends we need to support the deprecated `--list=hidden-options` but it should now be the same as just saying `--help`.

File: src/options.c
Comment (magnumripper) [inline src/options.c]: Still one `exit()` missing and that is why you suddenly got the 'Unknown option: "--help"'.

You need one `exit(0);` after line 486 here, that is INSIDE the `if (argc == 2 && ...)` block but OUTSIDE the `if (john_main_process)` block.

**Review by magnumripper — CHANGES_REQUESTED** [review]

**Comment by magnumripper** [comment]
> One detail to fix before merging: the "unknown option" error message.
> 
> ```
> $ ../run/john --help
> John the Ripper 1.9.0-jumbo-1+bleeding-4a413af 2020-10-28 23:54:06 +0300 OMP [linux-gnu 64-bit x86_64 AVX AC]
> Copyright (c) 1996-2020 by Solar Designer and others
> Homepage: https://www.openwall.com/john/
> 
> Usage: john [OPTIONS] [PASSWORD-FILES]
> Use --help to list all available options.
> 
> --help                     print usage summary
> [...]
> --format=[NAME|CLASS][,..] force hash of type NAME. The supported formats can
>                            be seen with --list=formats and --list=subformats. Valid
>                            classes: dynamic, cpu, opencl, ztex, mask, omp, all,
>                            enabled, disabled.
> Unknown option: "--help"
> ```

There's another minor problem with the above. The line "Use --help to list all available options." should only be printed for `./john` without options, not for `./john --help`.

File: src/options.c
Comment (magnumripper) [reply src/options.c]: BTW while at it, there's a space missing:
```diff
- 	} else if(argc < 2) {
+ 	} else if (argc < 2) {
```

**Comment by magnumripper** [comment]
> > @wsandin , IMO, if you create a new PR to handle shell completion, @magnumripper can approve the PR easier (he already tested the `--help` stuff).

True, but time isn't super critical so if you can get them to work within a day or two, just keep it all together.

> That's a good point. I saw your message just now, I'll rebase these out of this branch then and move to a separate PR after a quick review tomorrow - want to go through the completions again.

If you face tougher problems with shell completion really slowing you down, let's move them to a separate PR if you like.

**Comment by wsandin** [comment]
Coming right up!

File: src/listconf.c
Comment (wsandin) [reply src/listconf.c:404]: So I've been looking into that and would like to ask for some feedback what you think would be the best way to go on about that?

What comes to mind is moving the PRINT_HELP macros to a help.h or print_help.h and create a print_help(); function that I import in options.c.

Comment (magnumripper) [reply src/listconf.c:404]: I'd keep them in `options.c` (so eg. `opt_help()`) but yes, an exported function that we can call from here.

File: src/options.c
Comment (magnumripper) [inline src/options.c]: This macro now ends with `\` continuation followed by a blank, which again is fine but a potential foot-gun. While fixing that, you could have this last string of JOHN_BANNER end with double `\n\n` instead of repeatedly following `opt_banner();` with manually printing the second LF.

**Review by magnumripper — APPROVED** [review]

File: src/options.c
Comment (magnumripper) [reply src/options.c]: I made the suggested changes directly in Github GUI. If bots buy them, I'll merge soon.

BTW I also inlined a single-line-print function only used once, and made another one static. I think that was all.

**Comment by wsandin** [comment]
I just went through your last commits and comments. :+1: 

Thanks @magnumripper :pray: for your patience and great feedback throughout. I will try to get started with the groff man pages and continue to look out for more issues.

**Comment by wsandin** [comment]
I noticed one small detail, basically a missing newline before the brackets in the methods. I'll fix quickly.

**Comment by magnumripper** [comment]
I accidentally dropped the reference to #4394 in the commit message. **Edit:** Oh wait, it should be #4340 so it would have been wrong anyway. Oh well. **Edit 2:** Somehow Github kept some reference anyway so the issue was auto-closed regardless.

Thanks @wsandin, we're hoping for more!

**Comment by solardiz** [comment]
Thank you @wsandin and @magnumripper for making this happen!

**Pull request merged**
